osu.Game/Overlays/BeatmapSet/Details.cs

```diff
@@ -101,12 +99,22 @@ public Details()
             };
         }
 
-        [BackgroundDependencyLoader]
-        private void load()
+        [Resolved]
+        private RulesetStore rulesets { get; set; }
+
+        protected override void LoadComplete()
         {
+            base.LoadComplete();
             updateDisplay();
         }
 
+        private void updateDisplay()
+        {
+            Ratings.Ratings = BeatmapSet?.Ratings;
+            ratingBox.Alpha = BeatmapSet?.Status > 0 ? 1 : 0;
+            advanced.Ruleset.Value = rulesets.GetRuleset(beatmapInfo?.Ruleset.OnlineID ?? 0);
```

Comment:
I dunno. I would have just changed `AdvancedStats` to have a ctor flag of something like "showConvertedRulesetInfo" rather than changing this whole flow.

The new flow reads very shoddy and is confusing. Like when do you hook up ruleset to beamap and when do you use game? It's not explained anywhere and bound to get used wrong in the future.

Reply:
I don't agree with that. I think boolean flags are shoddy and should only be used as a last resort. This seems better because it makes this a more "pure" component in that it has explicit inputs that it's reading rather some implicit resolution process as previously.

On a quick skim the only thing that makes me wince in this diff is the `SongSelect` change and it looks like that one is caused by the typical song select debounce hell.

Reply:
To confirm you're understand this: `Ruleset` is not set in the online beatmap set overlay, which is the "fix" here?

Zero xmldoc explaining this, just a wild ruleset bindable? I don't know.

Reply:
Uhhhh... I guess I missed that part. `null` fixing it is dodgy indeed.

Should be explicitly set in all instances.

Reply:
If it's explicitly set on the beatmap overlay to anything that isn't null, it will regress the fix. I'm open to a better structure or xmldoc trying to explain this, but this is why I proposed the ctor argument instead (it's explicit and it gives us a place to document the modes of operation). 🤷

Reply:
> If it's explicitly set on the beatmap overlay to anything that isn't null, it will regress the fix

...what

surely not? surely setting a non-mania ruleset in the beatmap set overlay will cause the key count thingy to stop displaying?

if not then the entire thing needs taking a torch to because that is absurd.

Reply:
What are you going to set it to? Just a re-reference to the beatmap's ruleset? That would work but is arguably weird to have to do.

Making ruleset required would kinda mean removing the fallback, but then you have two bindables which are cross-dependent.

A better api would be something like `ShowBeatmap(BeatmapInfo, RulesetInfo? convertRuleset = null)` but it would mean we can't use binding and would have to call it from song select on each user ruleset change.

Reply:
> What are you going to set it to? Just a re-reference to the beatmap's ruleset?

Yes. Clearly we need it at this stage. I don't think it's weird if converts exist.

> A better api would be something like `ShowBeatmap(BeatmapInfo, RulesetInfo? convertRuleset = null)`

Sure I can agree with that.

> would have to call it from song select on each user ruleset change

Why? `ShowBeatmap()` is a one-shot operation. The beatmap set overlay has a separate ruleset selector. It doesn't follow the game-global ruleset already.

Reply:
My API proposal is for `AdvancedStats` which is used at song select, and bound to song select's local `ruleset`.

Maybe if we can't easily agree on this the class should just be copypasted into two versions.

Reply:
I'm not sure why this conversation is going in the places it's going so I might just recuse myself from it.

> the next issue is going to be that it's reading from global mods and displaying "unexpected" on the overlay

I never wanted that. It should not be reading from global mods. It should be given the correct ruleset to show explicitly in every usage. No inferences.

Reply:
There's more issues with `AdvancedStats` ruleset – it doesn't have knowledge of whether a conversion can happen, so right now it will prefer the bound ruleset even if the conversion is impossible (ie. mania beatmap and osu! ruleset being set). Not sure if this is okay or not. Should it be up to the point of construction to get this right?

Reply:
My vision for it was for it to be a dumb component that displays what it's told to. At the point it's told to display an impossible conversion something has gone deeply wrong somewhere higher above I'd say.

Reply:
Fair enough.